base-notebook/start.sh

```diff
@@ -11,95 +12,169 @@ else
+            groupadd --gid $NB_GID --non-unique ${NB_GROUP:-${NB_USER}}
         fi
+        # Recreate the user as we want it
         userdel $NB_USER
-        useradd --home /home/$NB_USER -u $NB_UID -g $NB_GID -G 100 -l $NB_USER
+        useradd --home /home/$NB_USER --uid $NB_UID --gid $NB_GID --groups 100 --no-log-init $NB_USER
     fi
 
-    # Enable sudo if requested
+    # Conditionally enable passwordless sudo usage for the jovyan user
     if [[ "$GRANT_SUDO" == "1" || "$GRANT_SUDO" == 'yes' ]]; then
-        echo "Granting $NB_USER sudo access and appending $CONDA_DIR/bin to sudo PATH"
-        echo "$NB_USER ALL=(ALL) NOPASSWD:ALL" > /etc/sudoers.d/notebook
+        echo "Granting $NB_USER passwordless sudo rights!"
+        echo "$NB_USER ALL=(ALL) NOPASSWD:ALL" > /etc/sudoers.d/added-by-start-script
     fi
 
-    # Add $CONDA_DIR/bin to sudo secure_path
-    sed -r "s#Defaults\s+secure_path=\"([^\"]+)\"#Defaults secure_path=\"\1:$CONDA_DIR/bin\"#" /etc/sudoers | grep secure_path > /etc/sudoers.d/path
+    # Ensure that the initial environment that this container is started with
+    # is preserved when we run transition from running as root to running as
+    # NB_USER. But, make an exception for environment variables explicitly
+    # listed in the JUPYTER_ENV_VARS_TO_UNSET environment variable, let these be
+    # unset before the transition.
+    #
+    # - We use the sudo command to execute the command as NB_USER. But, what
+    #   happens to the environment will be determined by configuration in
+    #   /etc/sudoers and /etc/sudoers.d/* as well as flags we pass to the sudo
+    #   command. The behavior can be inspected with `sudo -V` run as root.
+    #
+    #   ref: `man sudo`    https://linux.die.net/man/8/sudo
+    #   ref: `man sudoers` https://www.sudo.ws/man/1.8.15/sudoers.man.html
+    #
+    # - We use the `--preserve-env` flag to pass through most environment, but
+    #   understand that exceptions are caused by the sudoers configuration:
+    #   `env_delete`, `env_check`, and `secure_path`.
+    #
+    # - We reduce the `env_delete` list of default variables to be deleted by
+    #   default which would ignore the `--preserve-env` flag and `env_keep`
+    #   configuration.
+    #
+    # - We manage the PATH variable specifically as `secure_path` is set by
+    #   default in /etc/sudoers and would override the PATH variable. So we
+    #   disable that default.
+    echo 'Defaults env_delete -= "PATH LD_* PYTHON*"' >> /etc/sudoers.d/added-by-start-script
+    echo 'Defaults !secure_path' >> /etc/sudoers.d/added-by-start-script
 
-    # Exec the command as NB_USER with the PATH and the rest of
-    # the environment preserved
+    # NOTE: This hook is run as the root user!
     run-hooks /usr/local/bin/before-notebook.d
-    echo "Executing the command: ${cmd[@]}"
-    exec sudo -E -H -u $NB_USER PATH=$PATH XDG_CACHE_HOME=/home/$NB_USER/.cache PYTHONPATH=${PYTHONPATH:-} "${cmd[@]}"
```

Comment:
> When the PATH=$PATH is here, it actually does the right thing and path is correct (/opt/conda/bin is in the front, not at the end).

I'm pretty sure that `PATH=$PATH` is not doing the right thing here, even though it may look like it is. See #1053 for details. It's very subtle, and even the `which` command gives the "wrong" answer.